### PR TITLE
JSON-RPC Client with FULLNODE_API_INFO config & JWT support

### LINKED ISSUE #1084: Forest CLI `FULLNODE_API_INFO` support, and --rpc_port override
**Issue Summary**
`lotus` can be run both in daemon mode, and also in client mode. The client allows HTTP JSON-RPC commands to be sent to a remote server. We should support that capability, also.

**ACs**
- [ ] `FULLNODE_API_INFO` environment variable, if present, is utilized
- [ ] `--rpc-port` argument can also be passed when invoking the forest client to override the port from either the default and FULLNODE_API_INFO environment variable

### PR DESCRIPTION
**Pull request by cryptoquick**
**Summary of changes**
Changes introduced in this pull request:
- Remove jsonrpsee from forest CLI
- Add surf for HTTP requests
- Parse FULLNODE_API_INFO env var to configure HTTP JSON-RPC client
- Pass Authorization header if JWT is provided

**Reference issue to close (if applicable)**

Closes #1084 and #998

**Other information and links**
Ran into a snag, so still WIP:
https://github.com/kardeiz/jsonrpc-v2/issues/17

Once I get the RPC client working, I'll refactor other RPC methods

The complexity here came from jsonrpsee not supporting:
- Dynamic variables in macro declarations
- Authorization headers

### CODE REVIEW

File: forest/src/cli/chain_cmd.rs
Comment (cryptoquick) [inline forest/src/cli/chain_cmd.rs:66]: @ec2 buddy, this didn't work like I thought it would
I want proper cids, don't I?

To compare, I ran:

`FULLNODE_API_INFO="/ip4/127.0.0.1/tcp/1234/http" lotus chain head && target/release/forest chain head`

An example of what this branch of forest outputs:

```
[
  "AXGg5AIgos5/lOLLPc4KgPo+FspWt6zhQg1jc9z5bMJhBBkVqVY=",
  "AXGg5AIgP3+0yylaV5kSH/H63wUHhgDku+R1ByR7SZAk1XFNcMc=",
  "AXGg5AIgCIQ7YlecRZgMjjN9mp3AqonaSsepSW/nvi9NicRj7i4=",
  "AXGg5AIg8YnMkQAyXr5nOevGu98pUKHpjDCW/jDfrPVpCYGMBEk=",
  "AXGg5AIgnaTQgw3872sqgeDsYiNkVetgLqqEKn834Ldf28eL+IE=",
  "AXGg5AIgqsaPhPKrKKlgnX1QB/W+Mc8JvpoqtAn+Wc+oAb0cBZw=",
  "AXGg5AIgZVYwSQSQv3wi//FxeDN/YJuVfnPNnmSQyO8VfwM2wzY=",
  "AXGg5AIgZdvSYybA9iJir15ZHycp+HsUIOuUjKIa803CoW63KM0=",
  "AXGg5AIgrmI5G4UujZZJF/5a4eIbCFKsRrH896+HmkIzifErRgE="
]
```

Meanwhile lotus outputs:

```
bafy2bzacecrm474u4lft3tqkqd5d4fwkk232zykcbvrxhxhzntbgcbazcwuvm
bafy2bzacea7x7nglffnfpgisd7y7vxyfa6dabzf34r2qojd3jgicjvlrjvymo
bafy2bzaceaeiio3ck6oelgamryzx3gu5ycvitwsky6uus37hxyxu3coempxc4
bafy2bzacedyytteraazf5pthhhv4no67ffikd2mmgclp4mg7vt2wscmbrqces
bafy2bzaceco2juedbx6o62zkqhqoyyrdmrk6wybovkccu7zx4c3v7w6hrp4ic
bafy2bzacecvmnd4e6kvsrklatv6vab7vxyy46cn6tivlicp6lhh2qan5dqczy
bafy2bzacebsvmmcjasil67bc77yxc6btp5qjxfl6opgz4zeqzdxrk7ydg3btm
bafy2bzacebs5xutde3apmitcv5pfshzhfh4hwfba5okiziq26ng4filow4um2
bafy2bzacecxgeoi3quxi3fsjc77fvypcdmefflcgwh6ppl4htjbdhcprfndac
```

Comment (cryptoquick) [reply forest/src/cli/chain_cmd.rs:65]: Eric let me know via PM last night that Cid implements `to_string()`, so that works now.

**Comment by cryptoquick** [comment]
Sprint carryover update:
I got a lot of good work done during last sprint on implementing the majority of what's needed to fulfill this task. 

** Done **

- RPC request building via jsonrpc_v2 (required collaboration with upstream dev on missing serialization trait, since jsonrpc_v2 isn't traditionally used for clients)
- Surf HTTP integration
- Auth header passing (untested)
- Multiaddress parsing via Regex (will update to an approach based on parity multisig as defined in this issue before I mark this as ready, though: https://github.com/libp2p/rust-libp2p/issues/2073 )
- Calling RPC functions without parameters
- Fixed serialization issues on Cids
- Fixed CI by bumping Circle rustc to 1.52.1 due to me, a chad rust dev, using freshly stable std lib methods

** Doing **

- Debugging RPC methods with parameters, which are currently broken due to a deserialization issue
- Test all forest CLI methods, with both forest and lotus nodes
- Test JWT creation and passing

File: node/rpc-client/src/client.rs
Comment (ec2) [inline node/rpc-client/src/client.rs:41]: Dont you need to check if its http? This regex i think only captures ip4 and tcp

Comment (cryptoquick) [reply node/rpc-client/src/client.rs:41]: Good point, it should work with the output from `lotus auth api-info`. I'll be working on that now in my multiaddress parsing refactor.

Comment (cryptoquick) [reply node/rpc-client/src/client.rs]: This works now:

`FULLNODE_API_INFO="/ip4/127.0.0.1/tcp/1234/http" lotus auth api-info --perm admin`

> FULLNODE_API_INFO=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.6dMalXxBGadVxVLzbxAanYV--7ihmIGK3eLrtjhEawQ:/ip4/127.0.0.1/tcp/1234/http

`FULLNODE_API_INFO=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.6dMalXxBGadVxVLzbxAanYV--7ihmIGK3eLrtjhEawQ:/ip4/127.0.0.1/tcp/1234/http target/debug/forest auth create-token -p write`

> 2021-05-28T16:25:29.142Z INFO  rpc_client::client > Using JSON-RPC v2 HTTP URL: http://127.0.0.1:1234/rpc/v0
> eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiXX0.KBtaQC9K8Tnu2tbyyf0pv0MMD8-kbA7cCVATLqS6F4o

**Review by connormullett — COMMENTED** [review]
Looks great, will approve once this is R4R if no other changes

**Review by connormullett — APPROVED** [review]
LGTM

**Comment by cryptoquick** [comment]
I was just writing this, but then I realized, we should use the `auth api-info` command instead. That's NYI, though.

```
### CLI Usage Examples

A tutorial for currently implemented `forest` client workflows.

> forest auth create-token -p admin

Take the JWT output and put it into multiaddress format: ugh n/m, let's wait till we have auth api-info
```

I'm down to merge this without the usage examples changes and the RPC client error parsing if everyone else agrees.

**Comment by connormullett** [comment]
> I was just writing this, but then I realized, we should use the `auth api-info` command instead. That's NYI, though.
> 
> ```
> ### CLI Usage Examples
> 
> A tutorial for currently implemented `forest` client workflows.
> 
> > forest auth create-token -p admin
> 
> Take the JWT output and put it into multiaddress format: ugh n/m, let's wait till we have auth api-info
> ```
> 
> I'm down to merge this without the usage examples changes and the RPC client error parsing if everyone else agrees.

I can add a note to create some usage examples with my work if needed. You and I understand the CLI enough to be able to work with it. Documentation and usage examples can come once we get a good number of commands running I feel

**Comment by cryptoquick** [comment]
@connormullett Yeah, makes sense. I'll make a separate issue for that work.

File: forest/src/cli/chain_cmd.rs
Comment (olibearo) [inline forest/src/cli/chain_cmd.rs:66]: In case you find it interesting: often times when running collect, the type inference engine just needs to know that you want to use the `Vec<_>` implementation, and can easily detect the `T` within the `Vec<_>`

**Review by ec2 — APPROVED** [review]
LGTM. I think it may potentially be useful for us to have the CLI be able to also use WS instead of HTTP in the future. Just a thought

File: forest/Cargo.toml
Comment (cryptoquick) [reply forest/Cargo.toml:32]: You're right, you did tell me to add it, tho. I'll take it out :)

**Review by olibearo — APPROVED** [review]
I only did a shallow code review since this already has two approvals, LGTM

**Comment by cryptoquick** [comment]
> LGTM. I think it may potentially be useful for us to have the CLI be able to also use WS instead of HTTP as an option in the future. Just a thought

I'll make a separate issue for that. It's an entire different transport, so it'd require changes to Multiaddress parsing and some abstraction for a WS client, in addition to surf for HTTP.

**Comment by cryptoquick** [comment]
> I only did a shallow code review since this already has two approvals, LGTM

ty :)

**Comment by ec2** [comment]
> > LGTM. I think it may potentially be useful for us to have the CLI be able to also use WS instead of HTTP as an option in the future. Just a thought
> 
> I'll make a separate issue for that. It's an entire different transport, so it'd require changes to Multiaddress parsing and some abstraction for a WS client, in addition to surf for HTTP.

Yes, technically, but since the HTTP and WS server are on the same port, you wont need to change the multiaddr parsing. But yeah there will be some heavy lifting done to abstract the client to handle http and ws. Just an idea i have :)

**Pull request merged**
